src/build.c

```diff
@@ -975,28 +965,23 @@ static GPid build_run_cmd(GeanyDocument *doc, guint cmdindex)
 	{
 		gchar *vte_cmd;
 
+		/* VTE expects commands in UTF-8 */
+		SETPTR(run_cmd, utils_get_utf8_from_locale(run_cmd));
+		SETPTR(working_dir, utils_get_utf8_from_locale(working_dir));
+
 		if (vc->skip_run_script)
-		{
-			SETPTR(vte_cmd_nonscript, utils_get_utf8_from_locale(vte_cmd_nonscript));
-			vte_cmd = g_strconcat(vte_cmd_nonscript, "\n", NULL);
-			g_free(vte_cmd_nonscript);
-		}
+			vte_cmd = g_strconcat(run_cmd, "\n", NULL);
 		else
-			vte_cmd = g_strconcat("\n/bin/sh ", RUN_SCRIPT_CMD, "\n", NULL);
+			vte_cmd = g_strconcat("\n/bin/sh ", run_cmd, "\n", NULL);
 
-		/* change into current directory if it is not done by default */
-		if (! vc->follow_path)
-		{
-			/* we need to convert the working_dir back to UTF-8 because the VTE expects it */
-			gchar *utf8_working_dir = utils_get_utf8_from_locale(working_dir);
-			vte_cwd(utf8_working_dir, TRUE);
-			g_free(utf8_working_dir);
-		}
+		vte_cwd(working_dir, TRUE);
 		if (! vte_send_cmd(vte_cmd))
 		{
 			ui_set_statusbar(FALSE,
 		_("Could not execute the file in the VTE because it probably contains a command."));
 			geany_debug("Could not execute the file in the VTE because it probably contains a command.");
+			if (!vc->skip_run_script)
+				g_unlink(run_cmd);
```

Comment:
the question was whether at this point we know the command has read the script.  vte_send_cmd() just writes to the child, so I could imagine either the VTE shell itself, or `/bin/sh` wouldn't yet have finished their handling of the file when you remove it.
Unless there is a real synchronization somewhere, I'm afraid this might be a race condition at best